Add configure SS and pipeline

mamedin opened the pull request:
This PR will use the features added by the following PRs:

-https://github.com/artefactual/archivematica/pull/681
-https://github.com/artefactual/archivematica-storage-service/pull/213

Specifically, this PR allows to:

a) Create a SS superuser
b) Create a dashboard admin user and register the pipeline on the SS.

To configure SS and pipeline on a deployment, the following boolean
extra-vars have to be enabled:

- archivematica_src_configure_dashboard : To create the dashboard admin user and
                             to register pipeline on SS
- archivematica_src_configure_ss : To create SS superuser

They are disabled by default.

connects to #161

mamedin committed:
Add configure SS and pipeline

This PR will use the features added by the following PRs:

-https://github.com/artefactual/archivematica/pull/681
-https://github.com/artefactual/archivematica-storage-service/pull/213

Specifically, this PR allows to:

a) Create a SS superuser
b) Create a dashboard admin user and register the pipeline on the SS.

To configure SS and pipeline on a deployment, the following boolean
extra-vars have to be enabled:

- archivematica_src_configure_dashboard: To create the dashboard admin user and
to register pipeline on SS
- archivematica_src_configure_ss: To create SS superuser

They are disabled by default.

mamedin commented:
This fixes #161

mamedin commented:
You can use the following variables on vars.yml file for testing:

archivematica_src_configure_ss_user: "ss_user"
archivematica_src_configure_ss_password: "password_ss"
archivematica_src_configure_ss_email: "test@example.com"
#archivematica_src_configure_ss_api_key: "9fd9a15258c244b2ca4bbee0a2393809"
archivematica_src_configure_ss_url: "http://SS_IP_ADDRESS:8000"
archivematica_src_configure_am_user: "dashboard_user"
archivematica_src_configure_am_password: "password_dashboard"
archivematica_src_configure_am_email: "test@example.com"
#archivematica_src_configure_am_api_key: "058d6ca988fecd5a6bf4ab24492e7ee8c954f485"
archivematica_src_configure_am_org_name: "org_name"
archivematica_src_configure_am_org_id: "12345"
#archivematica_src_configure_am_whitelist: '"1.2.3.3 127.0.0.1"'

To run the playbook with the configure feature, use the extravars: 

- "archivematica_src_configure_ss_dashboard=yes": To configure SS and dashboard
- "archivematica_src_configure_ss=yes": To create SS superadmin only
- "archivematica_src_configure_dashboard=yes": To create the dashboard andmin and register the pipeline.

mamedin commented:
I forgot to mention the more important thing. There's a breaking change in the PR https://github.com/artefactual/archivematica/pull/681 about the dashboard API whitelist: an empty whitelist means: "allow from all". 

Do you think that a better default whitelist could be "localhost" or another value?

scollazo commented on defaults/main.yml:
```diff
@@ -116,3 +121,16 @@ archivematica_src_automationtools_transfers_pidfile: "/var/archivematica/automat
+archivematica_src_configure_am_password: "test"              # Dashboard admin password
+archivematica_src_configure_am_email: "test@example.com"     # Dashboard admin password
+archivematica_src_configure_am_org_name: "test"              # Dashboard admin Org name
+archivematica_src_configure_am_org_id: "test"                # Dashboard admin Org id
```
Can you add also the "First name" and "last name" fields? This way, we have all the configuration options exposed here.

scollazo commented on tasks/pipeline-configure.yml:
```diff
@@ -0,0 +1,14 @@
+
+- name: "Create Dashboard user and register pipeline on SS"
+  django_manage:
+    command: "install --username={{ archivematica_src_configure_am_user }} --password={{ archivematica_src_configure_am_password }} --email={{ archivematica_src_configure_am_email }} --org-name={{ archivematica_src_configure_am_org_name }} --org-id={{ archivematica_src_configure_am_org_id }} --api-key={{ archivematica_src_configure_am_api_key }} --ss-url={{ archivematica_src_configure_ss_url }} - ...
```
This should be one line for each switch, it would improve readability

scollazo commented on tasks/ss-configure.yml:
```diff
@@ -0,0 +1,14 @@
+- set_fact: archivematica_src_configure_ss_api_key={{ 999999999999999999999 | random | to_uuid | hash('md5') }}
+  when: archivematica_src_configure_ss_api_key is undefined
+
+- name: "Create SS superuser"
```
What do you think about deleting the default test user when this is set?

scollazo commented on tasks/pipeline-configure.yml:
```diff
@@ -0,0 +1,14 @@
+- set_fact: archivematica_src_configure_am_api_key={{ 999999999999999999998 | random | to_uuid | hash('md5') }}
+  when: archivematica_src_configure_am_api_key is undefined
+
+- name: "Create Dashboard user and register pipeline on SS"
```
You have `archivematica_src_configure_dashboard` and `archivematica_src_configure_ss_dashboard: "no"`, but seems  like this configures both things always.

scollazo commented on defaults/main.yml:
```diff
@@ -116,3 +121,16 @@ archivematica_src_automationtools_transfers_pidfile: "/var/archivematica/automat
+# Configure AM
+# archivematica_src_configure_am_api_key & archivematica_src_configure_ss_api_key can be defined as vars. If not defined they will be autogenerated by the playbook
+archivematica_src_configure_ss_user: "test"                  # SS superuser
+archivematica_src_configure_ss_password: "test"              # SS superuser password
```
I think passwords should be something line "PleaseChangeMe!", so users don't forget to set them

scollazo commented on tasks/main.yml:
```diff
@@ -159,3 +159,23 @@
+# Configure SS
+#
+
+- include: "ss-configure.yml"
```
As the files have only two tasks each, what about consolidating them into  a single one?

scollazo reviewed (changes requested):
Some feedback provided

mamedin commented:
Thanks Santi, I have considered all your suggestions except for entering the variables "First name" and "Last name"

The variables "First name" and "Last name" were not contemplated in the https://github.com/artefactual/archivematica/pull/681 PR, so it requires an AM additional change.

I have simplified the variables and tags since it could lead to confusion and it made the configuration more complicated. Now there is only the `amsrc-configure` tag and the variables ``archivematica_src_configure_ss and `archivematica_src_configure_dashboard`.

The default SS test user is deleted when adding a custom SS superuser.

I have consolidated all task into a single file, called configure.yml.

The default password was set to "PleaseChangeMe!"

The format is now cleaner introducing some line breaks to facilitate the reading of the code.

This PR is ready for CR again.

scollazo commented on tasks/configure.yml:
```diff
@@ -0,0 +1,67 @@
+  become: "yes"
+  shell: |
+    cd {{ archviematica_src_ss_app }}
+    export SS_DB_NAME={{ archivematica_src_ss_environment['SS_DB_NAME'] }}
```
I think you can set the env in a line, as we do here: https://github.com/artefactual-labs/ansible-archivematica-src/blob/a38b2d1a8a972f51f395cffb6cf52ef2155b291d/tasks/ss-db.yml#L17

scollazo commented on tasks/configure.yml:
```diff
@@ -0,0 +1,67 @@
+    executable: /bin/bash
+  when: "archivematica_src_configure_ss|bool"
+
+#source {{ archivematica_src_ss_virtualenv }}/bin/activate;
```
Is this dead code?

scollazo reviewed (changes requested):
Some more comments

scollazo commented:
I'm thinking about adding a "security" measure, to avoid breaking currently configured systems:

- Check if there is an "admin" user created
- If so, trigger an ansible prompt and ask if we want to continue.

sevein commented:
@mamedin 🏓

sevein reviewed (commented):
Let's try to get this merged? It's a nice feature.

mamedin commented:
Hi @sevein to merge:

-I need to change the PR to taking account the last @scollazo's suggestions:
        * Remove dead code
        * Set one-line environment

- Import these Jisc commits to Archivematica and Storage Service (to make idempotent):
        * https://github.com/JiscRDSS/archivematica/pull/46
        * https://github.com/JiscRDSS/archivematica/pull/55
        * https://github.com/JiscRDSS/archivematica-storage-service/pull/19/files

sevein commented:
@mamedin ops ok I see. Any of that is going to happen for AM17 so I guess we can ignore this PR for now. Thanks!

scollazo commented:
Although not idempotent,  the changes in this PR work on am 1.7. @mamedin can you address my comments about setting environment vars so we can merge this?

We should also create a bug in archivematica in order to get the changes needed for idempotency into 1.7.1, or at least in qa/1.x

mamedin commented:
@scollazo It is ready for QA, I have included your suggestions about stting environment.

scollazo commented on defaults/main.yml:
```diff
@@ -16,6 +16,14 @@ archivematica_src_install_acceptance_tests: "no"
+archivematica_src_configure_dashboard: "no"
+archivematica_src_configure_ss: "no"
+
+#Components to configure
```
Duplicated?

scollazo commented on defaults/main.yml:
```diff
@@ -177,3 +185,15 @@ archivematica_src_fixity_ss_url: "http://localhost:8000"
+archivematica_src_configure_am_user: "test"                  # Dashboard admin
+archivematica_src_configure_am_password: "PleaseChangeMe!"   # Dashboard admin password
+archivematica_src_configure_am_email: "test@example.com"     # Dashboard admin password
+archivematica_src_configure_am_org_name: "test"              # Dashboard admin Org name
```
Use "test org" for org, and "test id" for id?

scollazo commented on tasks/configure.yml:
```diff
@@ -0,0 +1,56 @@
+
+# Create api-key if not defined
+- set_fact: archivematica_src_configure_am_api_key={{ 999999999999999999998 | random | to_uuid | hash('md5') }}
+  when: archivematica_src_configure_am_api_key is undefined
```
For the ss api key, the condition is ` when: "archivematica_src_configure_ss_api_key is undefined and archivematica_src_configure_ss|bool"`, but here you are not checking for `archivematica_src_configure_dashboard`

mamedin replied:
Yes, it was done intentionally in case just SS is installed.

scollazo replied:
I'm not getting your point, I think that the other case (deploying the dashboard, with the ss in other server), will also fail.

Thinking out loud , does it make sense to place this steps in pipeline-dbconf.yml for the dashboard, and ss-db.yml for the storage service? This way you can rely on `archivematica_src_install_am` and `archivematica_src_install_ss` to avoid this issue.

scollazo reviewed (changes requested):
Feedback added. What happens if you run this playbook twice? Does the password get reset?

scollazo commented on tasks/configure.yml:
```diff
@@ -0,0 +1,56 @@
+- name: "Delete default SS superuser"
+  become: "yes"
+  shell: |
+    cd {{ archviematica_src_ss_app }}
```
I think that using `cd`, and `chdir` later, is not needed

scollazo replied:
Also, typo

mamedin replied:
Yes, this "cd" is not needed. The typo is prior to this change, and this variable is used several times in the role. I'll include the typo fix as a separate commit.

mamedin commented:
Pending to the https://github.com/artefactual/archivematica-storage-service/pull/364#issuecomment-395894237

I found some problems that I want to fix:
- Deleting the SS test user could be a problem when running this role several times
- It fails using the `amsrc-configure` tag when the `archivematica_src_mcpclient_clamav_tcp_port` variable is not defined

sevein commented:
Related: https://github.com/artefactual/archivematica/issues/1131.

sevein commented:
Related: https://github.com/artefactual/archivematica/issues/1134.

mamedin committed:
Add configure pipeline checks

This commit adds two test to check whether the dashboard user
`archivematica_src_configure_am_user` exists and the pipeline is already
registered before running the configure tasks. These new tasks will abort the
playbook when any of these basic test fails.

mamedin commented:
I have updated again this PR. I added 2 basic test looking at the dashboard database to check whether the pipeline is already registered or the dashboard user already exists. With these 2 basic tests we make sure that this role only runs in new deployments, which is where it makes sense.

I have rebased again to check the last commits and now there are no problems when using the `amsrc-configure` tag and some variables are not defined.

It is ready for CR again.

scollazo commented:
Instead of aborting if the user is already created, why not check for it first, and only create it when needed?

mamedin commented:
Hi @scollazo , the https://github.com/artefactual/archivematica/pull/1165 PR will make the dashboard installer idempotent, so the last commit could be removed because it is not needed.

jhsimpson commented:
I made a copy of this branch and rebased my copy on top of the recent changes in the master branch of this repo: https://github.com/jhsimpson/ansible-archivematica-src/tree/dev/configure-ss-dashboard

I tested this deploying to an Ubunt 16.04 host, using this dev branch of deploy-pub:

https://github.com/artefactual/deploy-pub/tree/dev/deploy-am-post17

This worked great for me, on a fresh install.  I haven't tested upgrades or other os'es.  I did have to play around a bit to figure out what ansible variables I actually needed.  You can see the variables I used here:https://github.com/artefactual/deploy-pub/blob/dev/deploy-am-post17/playbooks/archivematica/vars-singlenode-qa.yml

mamedin committed:
configure: remove abort on fails

mamedin commented:
In this last commit I have changed a bit the order of the tasks and checks, and now instead of abort when it detects the pipeline was configured or the pipeline user exists, it does nothing and sends a message.

mamedin committed:
configure: add environment to register pipeline

scollazo commented:
It has been a long way, but imho, this is ready for merge, kudos @mamedin !

We should add those new values to the docs in a separate commit.

scollazo reviewed (approved):
LGTM!
